solutions/usaco-965.mdx

```diff
@@ -2,90 +2,272 @@
-
-		// convert names to indices
-		int c1 = index(cow1);
-		int c2 = index(cow2);
-
-		// Add to Adjecency List
-		adj[c1].push_back(c2);
-		adj[c2].push_back(c1);
+	std::ifstream read("lineup.in");
+	int req_num;
+	read >> req_num;
+	vector<vector<int>> neighbors(COWS.size());
+	for (int r = 0; r < req_num; r++) {
+		string cow1;
+		string cow2;
+		string trash;
+		read >> cow1 >> trash >> trash >> trash >> trash >> cow2;
+
+		// Convert the names to their index in the list
+		int c1 = index(COWS, cow1);
+		int c2 = index(COWS, cow2);
+		neighbors[c1].push_back(c2);
+		neighbors[c2].push_back(c1);
 	}
 
-	vector<int> seq;  // sequence of cow indices
-	vector<bool> done(8);
-	for (int i = 0; i < 8; i++) {
-		if (!done[i] && adj[i].size() <= 1) {
-			// add path to sequence
-			int cur = i;
-			while (adj[cur].size() == 1) {
-				seq.push_back(cur);
-				done[cur] = 1;
-				int nex = adj[cur][0];
-				adj[nex].erase(find(begin(adj[nex]), end(adj[nex]), cur));
-				cur = nex;
+	vector<int> order;
+	vector<bool> added(COWS.size());
+	for (int c = 0; c < COWS.size(); c++) {
+		/*
+		 * Check that:
+		 * 1. This cow hasn't already been added yet.
+		 * 2. This cow could be the possible start of a chain.
+		 */
+		if (!added[c] && neighbors[c].size() <= 1) {
+			added[c] = true;
+			order.push_back(c);
+
+			// If the chain length > 1, we keep on going
+			if (neighbors[c].size() == 1) {
+				int prev = c;
+				int at = neighbors[c][0];
+				while (neighbors[at].size() == 2) {
+					added[at] = true;
+					order.push_back(at);
+					
+					int a = neighbors[at][0];
+					int b = neighbors[at][1];
+					int temp_at = a == prev ? b : a;
+					prev = at;
+					at = temp_at;
+				}
+				// Add the final element
+				added[at] = true;
+				order.push_back(at);
 			}
-			seq.push_back(cur);
-			done[cur] = 1;
 		}
 	}
 
-	// Print Output
-	for (int i = 0; i < seq.size(); i++) {
-		int t = seq[i];
-		cout << cows[t] << '\n';
+	std::ofstream out("lineup.out");
+	for (int c : order) {
+		out << COWS[c] << endl;
+	}
+}
+```
+
+</CPPSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class LineUp {
+	// Assumed to be in sorted order (which it is)
+	static final List<String> COWS = new ArrayList<>(
+		Arrays.asList(new String[] {
+			"Beatrice",
+			"Belinda",
+			"Bella",
+			"Bessie",
+			"Betsy",
+			"Blue",
+			"Buttercup",
+			"Sue"  
+		})
+	);
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader read = new BufferedReader(new FileReader("lineup.in"));
+		int reqNum = Integer.parseInt(read.readLine());
+
+		List<Integer>[] neighbors = new ArrayList[COWS.size()];
+		for (int c = 0; c < COWS.size(); c++) {
+			neighbors[c] = new ArrayList<>();
+		}
+		for (int r = 0; r < reqNum; r++) {
+			String[] words = read.readLine().split(" ");
+
+			// Convert the names to their index in the list
+			int cow1 = COWS.indexOf(words[0]);
+			int cow2 = COWS.indexOf(words[words.length - 1]);
```

Comment:
this makes the solution O(n^2)

Reply:
<img width="429" alt="image" src="https://user-images.githubusercontent.com/55369003/154201107-fc1b4ac7-6edb-47db-9104-cb3b2f3b74d5.png">
well, benq's solution does this as well
but ok i'll uh use a hashmap

Reply:
fixed